include/envoy/network/dns.h

```diff
@@ -45,12 +45,17 @@ class DnsResolver {
 public:
   virtual ~DnsResolver() = default;
 
+  /**
+   * Final status for a DNS resolution.
+   */
+  enum class ResolutionStatus { Success, Failure };
+
   /**
    * Called when a resolution attempt is complete.
-   * @param response supplies the list of resolved IP addresses and TTLs. The list will be empty if
```

Comment:
> The list will be empty if the resolution failed.

Most of the callback targets actually use this implicit "failure" hint. The strict dns cluster does not, which is why #9927 happens.

I am perusing through c-ares and trying to assert if according to RFC we can reasonably expect ARES_SUCCESS with no resolved addresses. @mattklein123 do you know off of the top of your mind?

If that is the case, then we don't _necessarily_ need to add the status. Just make sure the callback targets behave as we expect when the address list is empty.

Reply:
Yes it is definitely possible for a DNS resolution to return nothing and be successful, so we definitely need to handle this differently. What we do today is really just a hack. If higher layer code *additionally* wants to stabilize hosts in the case of 0 legitimate responses that is fine (e.g. I think logical DNS should do this, but strict DNS should not).

Reply:
Yep, agreed it is a hack. But I wanted to assert if it was a hack based on actual possible behavior; c-ares would only have an empty list with non-ARES_SUCCESS status. It wasn't immediately clear to me from my reading. Given that is not the case, I'll go ahead and continue with this change. Thanks!